### ISSUE TITLE
Bookmarks

### ISSUE THREAD

**Issue by flyasky** [xournalpp/xournalpp#846]
Can you implement bookmarks for pdf documents, please? It will be useful for fast navigation through big documents.

**Comment by andreasb242** [comment]
Already implemented. All PDF Bookmarks are displayed in the Sidebar.

**Comment by flyasky** [comment]
how to add new one into opened document wnen I drow over it?

**Comment by LittleHuba** [comment]
Are we talking about bookmarks included in the PDF or bookmarks you want to add manually in Xournal++ to a document you create?

**Comment by flyasky** [comment]
about document I created in xournal++

**Comment by LittleHuba** [comment]
This is about creating new bookmarks in Xournal++ not viewing bookmarks of PDF files

**Comment by coldspark29** [comment]
I would like to create my personal bookmarks as well, so +1 on this

**Comment by coldspark29** [comment]
Also the scrolling of the bookmarks is not possible with touch. Make an own issue for it?

**Comment by LittleHuba** [comment]
@coldspark29 please make an own issue for that. It keeps things easier to manage.

**Comment by onecan** [comment]
[Stylus Labs](http://www.styluslabs.com/) Write has an implementation of this feature that is quite nice to use. [Here](https://youtu.be/eWyxTKa2tZ0?t=144) is a video demo.

**Comment by fgheng** [comment]
+1 on this

**Comment by LittleHuba** [comment]
This will be included in the new fileformat.

**Comment by Alphaoo1** [comment]
Is this implemented yet?
How do I use it?

**Comment by moonsh4ke** [comment]
Any news?

**Comment by phyelsec** [comment]
How could I add that functionality, do I have to study the whole source code?
Sorry, I'm new contributing to open-source projects.

**Comment by LittleHuba** [comment]
You will have to at least get familiar with part of the codebase. The document model and rendering is the most important part for this work.

**Comment by grifball** [comment]
I implemented this very poorly, so:
BE CAREFUL WITH FILES THAT YOU CARE ABOUT
I added a button to toggle bookmarks on pages as well as a pair of buttons to skip between bookmarks (similar to next/prev page, but skipping all non-bookmarked pages).
I then updated the sidebar to show bookmarked pages with a green border and modified the save/load code to preserve bookmarks. The save/load code is what I'm most worried about if you try to use this.

Eventually, there should be a graphic that agrees with the selection graphic (the red border) such as an image of a bookmark. Also, the icons I made for the button should probably be remade. I might eventually fix this up if I find more motivation to work on xournalpp, but maybe this initial code could help someone else finish this feature as well.

https://github.com/grifball/xournalpp/commits/bookmarks

https://scottgriffy.com/xournalpp_bookmarks.webm

**Comment by algopenne** [comment]
+1, grifball's implementation looks good

**Comment by CodingByMe** [comment]
Is it possible to add bookmarks to a document now?

**Comment by grifball** [comment]
> Is it possible to add bookmarks to a document now?
Yeah, I used this branch for a year without it corrupting any data so I think it's fine. It should be rewritten to integrate with the existing bookmarks though (i.e. it should add bookmarks to the chapter list which is normally read from the pdf). I've switched to developing on xournalpp-mobile for Android ( https://gitlab.com/grifball/xournalpp_mobile ) so I might not implement this for a while.

**Comment by CodingByMe** [comment]
> @CodingByMe
> 
> > Is it possible to add bookmarks to a document now?
> 
> Yeah, I used this branch for a year without it corrupting any data so I think it's fine. It should be rewritten to integrate with the existing bookmarks though (i.e. it should add bookmarks to the chapter list which is normally read from the pdf). I've switched to developing on xournalpp-mobile for Android ( https://gitlab.com/grifball/xournalpp_mobile ) so I might not implement this for a while.

@grifball Thank you so much for answering me, I use xournal++ on Windows 10, can I still install the branch you created?

**Comment by grifball** [comment]
> > @CodingByMe
> > > Is it possible to add bookmarks to a document now?
> > 
> > 
> > Yeah, I used this branch for a year without it corrupting any data so I think it's fine. It should be rewritten to integrate with the existing bookmarks though (i.e. it should add bookmarks to the chapter list which is normally read from the pdf). I've switched to developing on xournalpp-mobile for Android ( https://gitlab.com/grifball/xournalpp_mobile ) so I might not implement this for a while.
> 
> @grifball Thank you so much for answering me, I use xournal++ on Windows 10, can I still install the branch you created?

@CodingByMe yeah I think compiling it on Windows would be possible as I didn't change the build system at all. I compiled it for Linux originally. Let me know how that pans out.

**Comment by Aitanuqui** [comment]
> I implemented this very poorly, so: BE CAREFUL WITH FILES THAT YOU CARE ABOUT I added a button to toggle bookmarks on pages as well as a pair of buttons to skip between bookmarks (similar to next/prev page, but skipping all non-bookmarked pages). I then updated the sidebar to show bookmarked pages with a green border and modified the save/load code to preserve bookmarks. The save/load code is what I'm most worried about if you try to use this.
> 
> Eventually, there should be a graphic that agrees with the selection graphic (the red border) such as an image of a bookmark in the corner of the sidebar page preview. Also, the icons I made for toggling/navigating bookmarks should probably be remade. I might eventually fix this up if I find more motivation to work on xournalpp, but maybe this initial code could help someone else finish this feature as well.
> 
> Oh, also, I didn't know that xournalpp actually did load "bookmarks" from a pdf, so I probably implemented this in a way that most people won't like (in the page preview rather than the bookmarks sidebar).
> 
> https://github.com/grifball/xournalpp/commits/bookmarks
> 
> https://scottgriffy.com/xournalpp_bookmarks.webm

I'm new to this niche. How do I install the implementation? Should I clone the code, or how is it done? I have installed the original Xournal++ app from the binaries downloaded on the website, but have no idea how to add your implementation.

**Comment by jereyes4** [comment]
As an alternative, I've written a plugin for bookmarks

https://github.com/jereyes4/xournalpp-bookmarks

It uses layers as bookmarks, and can export them to PDF too.

**Comment by Siddhnak** [comment]
Hi @jereyes4 

thanks for sharing your plugin.

i am facing some trouble using it..how to create bookmarks?
Will i be able to see any pop up?

i have followed the steps mentioned [here ](https://github.com/xournalpp/xournalpp/discussions/4522#discussioncomment-8789465) installed the lua-rocks directly from the installer as it was available for windows but i can't still use the plugin.
and can see the error you need to have Lua lgi-module when trying to view bookmarks although didn't create any XD.
![image](https://github.com/user-attachments/assets/5122abdc-b7ff-4971-a397-7f83edcb10fe)

**Comment by rolandlo** [comment]
@Siddhnak If you install the latest [nightly build](https://github.com/xournalpp/xournalpp/releases/tag/nightly) of Xournal++ the Lua `lgi`-module should work out of the box (on Windows).

**Comment by Siddhnak** [comment]
@rolandlo 
Thanks for replying back..

Nightly build
 works like a charm..

**Comment by jereyes4** [comment]
Ah I'm so sorry for the late reply. The plugin should include custom buttons that can be added to the toolbar, if that's what you mean.

**Comment by TamadoIchikai** [comment]
Hello, i just want to share my [fork](https://github.com/TamadoIchikai/xournalpp-bookmarks) of [jereyes4 plugins](https://github.com/jereyes4/xournalpp-bookmarks), some changes include:
- Shortcuts `ctrl+b` and `ctrl+shift+b` to create bookmark and open bookmarks manager.
- Bookmark layer now alway sorted bellow the drawing layer
- Double click to Jump to bookmarks and Double click on the bookmark name to edit.
- Bookmarks in the same page will be sorted alphabetically.
